File: src/wavefront/mtl.cpp
#include "mtl.hpp"

#include "wavefront/parse.hpp"

#include <exception>
#include <fstream>
#include <ostream>

using namespace std::experimental::filesystem;
using namespace glm;
using namespace std;

namespace trace {
namespace wavefront {
namespace {
// Mtl tokens
const char TOKEN_MTL_DIFFUSE[] = "kd";
const char TOKEN_MTL_DIFFUSE_MAP[] = "map_kd";
const char TOKEN_MTL_EMITTANCE[] = "emittance";
const char TOKEN_MTL_IOR[] = "indexofrefraction";
const char TOKEN_MTL_REFLECT0[] = "reflat0deg";
const char TOKEN_MTL_REFLECT90[] = "reflat90deg";
const char TOKEN_MTL_ROUGHNESS[] = "specularroughness";
const char TOKEN_MTL_SPECULAR[] = "ks";
const char TOKEN_MTL_TRANSPARANCY[] = "transparency";

// Light tokens
const char TOKEN_LIGHT_COLOR[] = "lightcolor";
const char TOKEN_LIGHT_INTENSITY[] = "lightintensity";
const char TOKEN_LIGHT_POSITION[] = "lightposition";
const char TOKEN_LIGHT_RADIUS[] = "lightradius";

// Camera tokens
const char TOKEN_CAMERA_FOV[] = "camerafov";
const char TOKEN_CAMERA_POSITION[] = "cameraposition";
const char TOKEN_CAMERA_TARGET[] = "cameratarget";
const char TOKEN_CAMERA_UP[] = "cameraup";
}

Mtl load_mtl(const path& file) {
  std::ifstream stream(file.string());
  if (!stream.good()) {
    string err = "Failed opening file '";
    err += file.string();
    err += "'";
    throw runtime_error(err);
  }

  Mtl mtl;

  string line;
  unsigned int line_index = 0;
  while (getline(stream, line)) {
    if (!line.empty()) {
      const char* token = skip_whitespace(line.c_str());

      if (equal("newmtl", token)) {
        mtl.materials.push_back({parse_string(skip_whitespace(token + 7)),
                                 "",
                                 {0.7f, 0.7f, 0.7f},
                                 {1.0f, 1.0f, 1.0f},
                                 {0.0f, 0.0f, 0.0f},
                                 0.001f,
                                 0.0f,
                                 0.0f,
                                 0.0f,
                                 1.0f});
      }

      else if (equal("newlight", token)) {
        mtl.lights.push_back(
            {{0.0f, 0.0f, 0.0f}, {1.0f, 1.0f, 1.0f}, 0.1f, 10.0f});
      }

      else if (equal("newcamera", token)) {
        mtl.cameras.push_back({{7.0f, 5.0f, 6.0f},
                               {0.0f, 0.0f, 0.0f},
                               {0.0f, 1.0f, 0.0f},
                               10.0f});
      }

#define TOKEN_VALUE(list, constant, parse, param, error) \
  else if (equal(constant, token)) {                     \
    list.back().param = parse(token + sizeof(constant)); \
  }

      TOKEN_VALUE(mtl.materials, TOKEN_MTL_DIFFUSE, parse_vec3, diffuse,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_DIFFUSE_MAP, parse_string,
                  diffuse_map, "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_EMITTANCE, parse_vec3, emittance,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_IOR, parse_float, ior,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_REFLECT0, parse_float, refl0,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_REFLECT90, parse_float, refl90,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_ROUGHNESS, parse_float, roughness,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_SPECULAR, parse_vec3, specular,
                  "No material created")
      TOKEN_VALUE(mtl.materials, TOKEN_MTL_TRANSPARANCY, parse_float,
                  transparency, "No material created")

      TOKEN_VALUE(mtl.lights, TOKEN_LIGHT_COLOR, parse_vec3, color,
                  "No light created")
      TOKEN_VALUE(mtl.lights, TOKEN_LIGHT_INTENSITY, parse_float, intensity,
                  "No light created")
      TOKEN_VALUE(mtl.lights, TOKEN_LIGHT_POSITION, parse_vec3, center,
                  "No light created")
      TOKEN_VALUE(mtl.lights, TOKEN_LIGHT_RADIUS, parse_float, radius,
                  "No light created")

      TOKEN_VALUE(mtl.cameras, TOKEN_CAMERA_FOV, parse_float, fov,
                  "No camera created")
      TOKEN_VALUE(mtl.cameras, TOKEN_CAMERA_POSITION, parse_vec3, position,
                  "No camera created")
      TOKEN_VALUE(mtl.cameras, TOKEN_CAMERA_TARGET, parse_vec3, target,
                  "No camera created")
      TOKEN_VALUE(mtl.cameras, TOKEN_CAMERA_UP, parse_vec3, up,
                  "No camera created")

#undef TOKEN_VALUE
    }

    ++line_index;
  }

  return mtl;
}
}
}

File: src/wavefront/mtl.rs
use nalgebra::{vector, Vector3};
use nom::IResult;
use nom::bytes::complete::tag_no_case;
use nom::character::complete::multispace0;
use nom::combinator::rest;
use nom::number::complete::float;
use nom::sequence::Tuple;

#[derive(Debug, Default, PartialEq)]
pub struct Material {
  pub name: String,
  pub diffuse_map: String,
  pub diffuse_reflection: Vector3<f32>,
  pub specular_reflection: Vector3<f32>,
  pub emittance: Vector3<f32>,
  pub transparency: f32,
  pub reflection_0_degrees: f32,
  pub reflection_90_degrees: f32,
  pub index_of_refraction: f32,
}

impl Material {
    pub fn new(name: String) -> Material {
        let mut material: Material = Default::default();
        material.name = name;
        material
    }
}

#[derive(Debug, Default, PartialEq)]
pub struct Light {
  pub position: Vector3<f32>,
  pub color: Vector3<f32>,
  pub radius: f32,
  pub intensity: f32,
}

#[derive(Debug, Default, PartialEq)]
pub struct Camera {
  pub position: Vector3<f32>,
  pub target: Vector3<f32>,
  pub up: Vector3<f32>,
  pub fov: f32,
}

#[derive(Debug, PartialEq)]
pub struct Mtl {
  pub materials: Vec<Material>,
  pub lights: Vec<Light>,
  pub cameras: Vec<Camera>,
}

fn tagged<'a, O>(name: &str, data: impl Fn(&'a str) -> IResult<&'a str, O>, input: &'a str) -> IResult<&'a str, O> {
    let (input, _) = tag_no_case(name)(input)?;
    let (input, _) = multispace0(input)?;
    data(input)
}

fn vec3(input: &str) -> IResult<&str, Vector3<f32>> {
    let (input, (x, _, y, _, z)) = (float, multispace0, float, multispace0, float).parse(input)?;
    Ok((input, vector![x, y, z]))
}

pub fn mtl(input: &str) -> Mtl {
  let mut materials: Vec<Material> = Vec::new();
  let mut lights: Vec<Light> = Vec::new();
  let mut cameras: Vec<Camera> = Vec::new();

  for line in input.lines() {
      let line = line.trim();
      if line.is_empty() || line.starts_with("#") {
          continue
      }

      if let Ok((_, _)) = tagged("newlight", rest, line) {
          lights.push(Default::default())
      } else if let Ok((_, x)) = tagged("lightposition", vec3, line) {
          lights.last_mut().unwrap().position = x
      } else if let Ok((_, x)) = tagged("lightcolor", vec3, line) {
          lights.last_mut().unwrap().color = x
      } else if let Ok((_, x)) = tagged("lightradius", float, line) {
          lights.last_mut().unwrap().radius = x
      } else if let Ok((_, x)) = tagged("lightintensity", float, line) {
          lights.last_mut().unwrap().intensity = x
      }

      else if let Ok((_, _)) = tagged("newcamera", rest, line) {
          cameras.push(Default::default())
      } else if let Ok((_, x)) = tagged("cameraposition", vec3, line) {
          cameras.last_mut().unwrap().position = x
      } else if let Ok((_, x)) = tagged("cameratarget", vec3, line) {
          cameras.last_mut().unwrap().target = x
      } else if let Ok((_, x)) = tagged("cameraup", vec3, line) {
          cameras.last_mut().unwrap().up = x
      } else if let Ok((_, x)) = tagged("camerafov", float, line) {
          cameras.last_mut().unwrap().fov = x
      }

      else if let Ok((_, name)) = tagged("newmtl", rest, line) {
          materials.push(Material::new(name.to_string()));
      } else if let Ok((_, x)) = tagged("kd", vec3, line) {
          materials.last_mut().unwrap().diffuse_reflection = x
      } else if let Ok((_, x)) = tagged("ks", vec3, line) {
          materials.last_mut().unwrap().specular_reflection = x
      } else if let Ok((_, x)) = tagged("reflat0deg", float, line) {
          materials.last_mut().unwrap().reflection_0_degrees = x
      } else if let Ok((_, x)) = tagged("reflat90deg", float, line) {
          materials.last_mut().unwrap().reflection_90_degrees = x
      } else if let Ok((_, x)) = tagged("indexofrefraction", float, line) {
          materials.last_mut().unwrap().index_of_refraction = x
      } else if let Ok((_, x)) = tagged("transparency", float, line) {
          materials.last_mut().unwrap().transparency = x
      } else if let Ok((_, _)) = tagged("specularroughness", float, line) {
          // TODO: not supported
      } else if let Ok((_, _)) = tagged("map_kd", rest, line) {
          // TODO: not supported
      }

      else {
          panic!("Unexpected line: \"{}\"", line)
      }
  }

  Mtl{materials, lights, cameras}
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_vec3() {
        assert_eq!(vec3("0 0 0"), Ok(("", vector![0., 0., 0.])));
        assert_eq!(vec3("1 2 3"), Ok(("", vector![1., 2., 3.])));
        assert_eq!(vec3("1. 2. 3."), Ok(("", vector![1., 2., 3.])));
        assert_eq!(vec3("-1. -2. -3."), Ok(("", vector![-1., -2., -3.])));
    }

    #[test]
    fn test_light() {
        assert_eq!(mtl("newlight l1").lights.len(), 1);
        assert_eq!(mtl("newlight l1\nlightposition 1. 2. 3.").lights[0].position, vector![1., 2., 3.]);
        assert_eq!(mtl("newlight l1\nlightcolor 1. 2. 3.").lights[0].color, vector![1., 2., 3.]);
        assert_eq!(mtl("newlight l1\nlightradius 1.").lights[0].radius, 1.);
        assert_eq!(mtl("newlight l1\nlightintensity 1.").lights[0].intensity, 1.);
    }

    #[test]
    fn test_camera() {
        assert_eq!(mtl("newcamera c1").cameras.len(), 1);
        assert_eq!(mtl("newcamera c1\ncameraposition 1. 2. 3.").cameras[0].position, vector![1., 2., 3.]);
        assert_eq!(mtl("newcamera c1\ncameratarget 1. 2. 3.").cameras[0].target, vector![1., 2., 3.]);
        assert_eq!(mtl("newcamera c1\ncameraup 1. 2. 3.").cameras[0].up, vector![1., 2., 3.]);
        assert_eq!(mtl("newcamera c1\ncamerafov 1.").cameras[0].fov, 1.);
    }

    #[test]
    fn test_mtl() {
        assert_eq!(mtl("newmtl m1").materials.len(), 1);
        assert_eq!(mtl("newmtl m1").materials[0].name, "m1");
        assert_eq!(mtl("newmtl m1\nkd 1. 2. 3.").materials[0].diffuse_reflection, vector![1., 2., 3.]);
        assert_eq!(mtl("newmtl m1\nks 1. 2. 3.").materials[0].specular_reflection, vector![1., 2., 3.]);
        assert_eq!(mtl("newmtl m1\nreflat0deg 1.").materials[0].reflection_0_degrees, 1.);
        assert_eq!(mtl("newmtl m1\nreflat90deg 1.").materials[0].reflection_90_degrees, 1.);
        assert_eq!(mtl("newmtl m1\nindexofrefraction 1.").materials[0].index_of_refraction, 1.);
        assert_eq!(mtl("newmtl m1\ntransparency 1.").materials[0].transparency, 1.);
        assert_eq!(mtl("newmtl m1\nspecularroughness 1.").materials.len(), 1);
        assert_eq!(mtl("newmtl m1\nmap_kd todo").materials.len(), 1);
    }
}
